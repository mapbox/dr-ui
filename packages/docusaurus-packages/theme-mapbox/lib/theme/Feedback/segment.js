import forwardEvent from './forward-event';
import * as Sentry from '@sentry/browser';
import env from '@theme/AnalyticsShell/env';

// Send the event to Segment with forward event
export function sendToSegment({ state, props }) {
  const event = createSegmentEvent({ state, props });
  forwardEvent(event, props.webhook, (err) => {
    if (err) Sentry.captureException(err);
  });
}

// Creates event to send to Segment
function createSegmentEvent({ state, props }) {
  const { user, anonymousId, feedback, sessionId, category, categoryType } =
    state;
  const { site, section, location } = props;
  // set user if available else set anonymousId (needed for forward-event request)
  const identity = {
    ...(user && user.id ? { userId: user.id } : { anonymousId: anonymousId })
  };

  return {
    event: 'Sent docs feedback',
    ...identity,
    properties: {
      // the unique id for the feedbak session
      sessionId,
      // the feedback category that the user selected (helpful for yes, unhelpful for no)
      category,
      // the sub type for the category
      categoryType,
      // text feedback, if available
      ...(feedback && { feedback: feedback }),
      // name of current site (important for filtering in Mode)
      site,
      // (optional) name of section for longer pagers, helpful for fitering in Mode and identifying section areas
      section,
      // get page context
      page: location || undefined,
      // set user if available else set anonymousId (needed for Mode)
      userId: user && user.id ? user.id : undefined,
      anonymousId: anonymousId,
      // set plan, if available
      ...(user && user.plan && { plan: user.plan.id }),
      // get environment: staging or production
      environment: env()
    }
  };
}
