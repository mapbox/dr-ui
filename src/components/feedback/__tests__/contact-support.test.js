import React from 'react';
import Feedback from '../feedback';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import forwardEvent from '../forward-event';

jest.mock('../forward-event');

describe('Workflow', () => {
  afterAll(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  const feedback = mount(
    <Feedback
      site="dr-ui"
      location={{
        pathname: '/mapbox-gl-js/api/',
        hash: '#lnglat'
      }}
      webhook={{
        production: '',
        staging: ''
      }}
    />
  );

  describe('1 - Open feedback', () => {
    test('Click button, then set state', () => {
      const btn = feedback.find('button');
      btn.simulate('click');
      expect(feedback.state().isOpen).toBeTruthy();
    });
    test('Send data to Segment', () => {
      expect(forwardEvent).toHaveBeenCalledWith(
        {
          anonymousId: expect.anything(),
          event: 'Sent docs feedback',
          properties: {
            anonymousId: expect.anything(),
            category: undefined,
            categoryType: undefined,
            contactSupport: false,
            environment: 'staging',
            exited: false,
            helpful: undefined,
            page: { hash: '#lnglat', pathname: '/mapbox-gl-js/api/' },
            section: undefined,
            sessionId: expect.anything(),
            site: 'dr-ui'
          }
        },
        { production: '', staging: '' },
        expect.any(Function)
      );
    });
  });

  describe('2 - Contact support', () => {
    test('Click button, then set state', () => {
      delete global.window.location;
      window.location = { assign: jest.fn() };
      const btn = feedback.find('button[value="Contact support"]');
      btn.simulate('click');
      expect(feedback.state().category).toBeUndefined();
      expect(feedback.state().contactSupport).toBeTruthy();
      expect(feedback.state().helpful).toBeUndefined();
      expect(toJson(feedback)).toMatchSnapshot();
      expect(window.location.assign).toBeCalledWith(
        'https://support.mapbox.com/'
      );
    });

    test('Then send to Segment', () => {
      expect(forwardEvent).toHaveBeenCalledWith(
        {
          anonymousId: expect.anything(),
          event: 'Sent docs feedback',
          properties: {
            anonymousId: expect.anything(),
            category: undefined,
            categoryType: undefined,
            contactSupport: true,
            environment: 'staging',
            exited: false,
            helpful: undefined,
            page: { hash: '#lnglat', pathname: '/mapbox-gl-js/api/' },
            section: undefined,
            sessionId: expect.anything(),
            site: 'dr-ui'
          }
        },
        { production: '', staging: '' },
        expect.any(Function)
      );
    });
  });
});
