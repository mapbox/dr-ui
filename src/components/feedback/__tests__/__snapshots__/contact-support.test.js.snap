// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workflow 2 - Contact support Click button, then set state 1`] = `
<Feedback
  feedbackSentryDsn="https://eccc8b561b9a461990309b01d33d54e3@sentry.io/1848287"
  location={
    Object {
      "hash": "#lnglat",
      "pathname": "/mapbox-gl-js/api/",
    }
  }
  site="dr-ui"
  type="page"
  webhook={
    Object {
      "production": "",
      "staging": "",
    }
  }
>
  <FeedbackWrapper
    title="Share your feedback"
  >
    <div
      className="dr-ui--feedback wmax300"
    >
      <div
        className="bg-gray-faint round py12 px12"
      >
        <div
          className="flex-parent flex-parent--column"
        >
          <div
            className="flex-child flex-parent flex-parent--space-between-main w-full mb12"
          >
            <div
              className="flex-child txt-bold"
            >
              Share your feedback
            </div>
            <div
              className="flex-child"
            >
              <Tooltip
                alignment="center"
                block={false}
                coloring="light"
                content="Close"
                disabled={false}
                maxWidth="medium"
                padding="small"
                placement="top"
                respondsToClick={false}
                textSize="s"
              >
                <PopoverTrigger
                  block={false}
                  content={
                    <div
                      className="txt-s wmax240"
                    >
                      Close
                    </div>
                  }
                  disabled={false}
                  popoverProps={
                    Object {
                      "alignment": "center",
                      "coloring": "light",
                      "hasPointer": true,
                      "hideWhenAnchorIsOffscreen": true,
                      "padding": "small",
                      "passthroughProps": Object {
                        "id": "tooltip-2",
                        "role": "tooltip",
                      },
                      "placement": "top",
                    }
                  }
                  receiveFocus={false}
                  renderHiddenContent={true}
                  respondsToClick={false}
                  respondsToFocus={true}
                  respondsToHover={true}
                  trapFocus={false}
                >
                  <div
                    className="inline-block"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                  >
                    <button
                      aria-describedby="tooltip-2"
                      aria-label="Close feedback"
                      className="link--gray"
                      id="feedback-close-button"
                      onClick={[Function]}
                    >
                      <Icon
                        inline={false}
                        name="close"
                        passthroughProps={Object {}}
                        size={18}
                      >
                        <svg
                          className="events-none icon"
                          focusable="false"
                          role="presentation"
                          style={
                            Object {
                              "height": 18,
                              "width": 18,
                            }
                          }
                        >
                          <use
                            xlinkHref="#icon-close"
                            xmlnsXlink="http://www.w3.org/1999/xlink"
                          />
                        </svg>
                      </Icon>
                    </button>
                    <div
                      className="hide-visually"
                      id="tooltip-2"
                      role="tooltip"
                    >
                      <div
                        className="txt-s wmax240"
                      >
                        Close
                      </div>
                    </div>
                  </div>
                </PopoverTrigger>
              </Tooltip>
            </div>
          </div>
          <div
            className="flex-child mb6 prose"
          >
            <button
              className="btn w-full mb6"
              key="I like this page"
              onClick={[Function]}
              value="I like this page"
            >
              I like this page
            </button>
            <button
              className="btn w-full mb6"
              key="Report a problem"
              onClick={[Function]}
              value="Report a problem"
            >
              Report a problem
            </button>
            <button
              className="btn w-full mb6"
              key="Something is confusing"
              onClick={[Function]}
              value="Something is confusing"
            >
              Something is confusing
            </button>
          </div>
          <div
            className="flex-child color-gray"
          >
            Need help?
             
            <button
              className="link"
              onClick={[Function]}
              value="Contact support"
            >
              Contact support
            </button>
          </div>
        </div>
      </div>
    </div>
  </FeedbackWrapper>
</Feedback>
`;
